Cython/Utility/FunctionArguments.c

```diff
@@ -422,7 +422,7 @@ static int __Pyx_MergeKeywords(PyObject *kwdict, PyObject *source_mapping) {
 #if CYTHON_METH_FASTCALL
     #define __Pyx_Arg_FASTCALL(args, i) args[i]
     #define __Pyx_NumKwargs_FASTCALL(kwds) PyTuple_GET_SIZE(kwds)
-    #define __Pyx_KwValues_FASTCALL(args, nargs) (&args[nargs])
+    #define __Pyx_KwValues_FASTCALL(args, nargs) ({ typeof (args) _a = (args); _a ? (&_a[nargs]) : NULL;})
```

Comment:
`typeof` isn't standard C so that's definitely not acceptable.

Is there any evidence that we actually call this with `args==NULL` and `nargs!=0`? I believe such a call wouldn't be following Python vectorcall protocol so would be dodgy anyway.

I do wonder if it should be `arg+nargs` though to avoid the dereference?

Reply:
> `typeof` isn't standard C

It [is](https://thephd.dev/ever-closer-c23-improvements#typeof), since C23. :)

More seriously, as the [paper](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2927.htm) that added it describes, this is a ubiquitous function (however it is spelled), because compilers need that functionality to implement `sizeof`.

Reply:
My mistake! It isn't part of the subset of the standards of C and C++ that Cython needs to be compatible with.

I also don't think it's hugely needed - I assume it's to avoid evaluating `arg` twice in the macro. But it's a macro that only Cython calls, and only with "simple" arguments, so that isn't a concern.

Reply:
> I do wonder if it should be arg+nargs though to avoid the dereference?

Even this is undefined behavior if `nargs != 0`.

> `({ typeof (args) _a = (args); _a ? (&_a[nargs]) : NULL;})`

This uses gcc statement expressions, which is a non standard extension of C/C++. This is not available e.g. on MSVC. As noted by @h-vetinari typeof is not officially available until C23, even though it is available in all compilers. So this is definetly not usable here.

> Is there any evidence that we actually call this with args==NULL and nargs!=0? I believe such a call wouldn't be following Python vectorcall protocol so would be dodgy anyway.

Yes, as long as we call `PyVectorcall_NARGS` this should never be the case. Since this got caught by UBSan I think we should find out how this part of the code could even be called with `args==NULL` and `nargs!=0`, since this sounds VERY wrong.